src/commonMain/kotlin/it/krzeminski/snakeyaml/engine/kmp/representer/BaseRepresenter.kt

```diff
@@ -232,36 +231,43 @@ abstract class BaseRepresenter(
     }
 }
 
-private class AnchorNodeMap : MutableMap<IdentityHashCode, Node> {
-
-    private val contents: MutableMap<IdentityHashCode, AnchorNode> = mutableMapOf()
-    private val contentsView: MutableMap<IdentityHashCode, Node>
-        get() = contents.mapValuesTo(mutableMapOf()) { (_, v) -> v.realNode }
-
-    override val size: Int get() = contents.size
-    override val entries: MutableSet<MutableMap.MutableEntry<IdentityHashCode, Node>> get() = contentsView.entries
-    override val keys: MutableSet<IdentityHashCode> get() = contentsView.keys
-    override val values: MutableCollection<Node> get() = contentsView.values
-
-    override fun clear() = contents.clear()
-    override fun containsKey(key: IdentityHashCode): Boolean = contents.containsKey(key)
-    override fun containsValue(value: Node): Boolean = contents.containsValue(value)
+/**
+ * The [IdentityLikeMap] tries to emulate `java.util.IdentityHashMap` from JVM world.
+ * However, considering limitations we have in KMP it makes the following assumption:
+ * **primitive types and strings does not have identity (meaning "test" and `new String("test")` would have the same "identity")**.
+ * This was made due to the following points on JVM:
+ * + [5.1.7. Boxing Conversion](https://docs.oracle.com/javase/specs/jls/se23/html/jls-5.html#jls-5.1.7) - primitives withing range -128..127 will have same identity hashcode
+ * + Two equal constant strings will have same identity hashcode. Only creating string explicitly using the constructor will result in a different identity in such case
+ *
+ * This is done because the main goal is to prevent the recursion and primitive types (and strings) cannot cause it.
+ */
+private open class IdentityLikeMap<T> private constructor(
+    private val contents: MutableMap<Any?, T>,
+) : MutableMap<Any?, T> by contents {
+    constructor() : this(HashMap())
 
-    override fun get(key: IdentityHashCode): Node? = contents[key]
+    override fun containsKey(key: Any?): Boolean = contents.containsKey(key.toIdentityKey())
 
-    override fun isEmpty(): Boolean = contents.isEmpty()
-    override fun put(key: IdentityHashCode, value: Node): Node? =
-        contents.put(key, AnchorNode(value))
+    override fun get(key: Any?): T? = contents[key.toIdentityKey()]
 
-    override fun putAll(from: Map<out IdentityHashCode, Node>): Unit =
-        contents.putAll(from.mapValues { (_, v) -> AnchorNode(v) })
+    override fun put(key: Any?, value: T): T? = contents.put(key.toIdentityKey(), value)
 
-    override fun remove(key: IdentityHashCode): Node? = contents.remove(key)
+    private fun Any?.toIdentityKey(): Any? = when (this) {
+        null,
+        is Byte,
+        is Short,
+        is Int,
+        is Long,
+        is Float,
+        is Double,
+        is Boolean,
+        is Char,
+        is String -> this
```

Comment:
Worth adding a comment that these types are primitive on at least one target platform (especially confusing for String).

Reply:
Good point, will do